feat(forms): introduce `ReadonlyFieldState` type for immutable field …

SkyZeroZx opened the pull request:
…state representation

Adds a `FormFieldSnapshot` type that provides a read-only view of the FormField, excluding methods that would mutate it.

Fixes #65779

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mmalerba reviewed (commented):
To quote myself from the associated issue:
> While I agree that there's no reason to be mutating the field state in this context, I'm not sure if its really worth the extra type complexity of introducing a ReadonlyFieldState. This does not feel like a common mistake people will be making

SkyZeroZx committed:
feat(forms): introduce `FormFieldSnapshot` type for immutable `FormField`

Adds a `FormFieldSnapshot` type that provides a read-only view of  `FormField`, excluding methods that would mutate it.

Fixes angular#65779
